README.md
# API Faker

使用YAML配置文件，快速创建一个Mock Server。

- 监听配置文件，自动刷新
- 根据响应文件的MIME类型设置`Content-Type`

## 安装

`go get -u github.com/fate-lovely/api-faker`

## 使用

```bash
$ api-faker --help
usage: api-faker [<flags>] <config>

Flags:
  -h, --help       Show context-sensitive help (also try
                   --help-long and --help-man).
      --port=3232  listening port
      --version    Show application version.

Args:
  <config>  config file
```

## 配置

配置文件使用YAML格式，每一个配置项由以下元素构成：

- `path`：指定请求路径
- `method`：指定请求方法，默认为`GET`
- `query`：指定请求应该携带的query参数
- `code`: 响应码，默认为200
- `headers`：自定义响应头
- `body`：响应体
- `file`：使用该文件内容作为响应体，如果此项存在，忽略`body`项，路径为绝对路径或者相对于配置文件的路径

对于每一个请求，会使用`path, method, query`进行匹配。

## 示例

**sample-config.yml**

```yaml
- path: /
  method: GET
  body: get /

- path: /
  method: POST
  body: post /

# 只有携带`lang=go`的查询字符串，才能匹配到这一条
- path: /query
  query:
    lang: go
  body: has lang=go query

- path: /headers
  headers:
    X-API-Faker: true
    X-Lang: Go
  body: respond with custom headers

# 使用`sample.json`进行响应，并自动设置MIME类型
- path: /json
  file: sample.json

# 响应码为201
- path: /created
  code: 201
  body: created
```

**sample.json**

```json
{
  "name": "api-faker",
  "msg": "hello world"
}
```

```bash
$ api-faker sample-config.yaml &
$ curl -i localhost:3232
HTTP/1.1 200 OK
Date: Wed, 29 Mar 2017 06:31:55 GMT
Content-Length: 5
Content-Type: text/plain; charset=utf-8

get /%
$ curl -i -XPOST localhost:3232/
HTTP/1.1 200 OK
Date: Wed, 29 Mar 2017 06:32:37 GMT
Content-Length: 6
Content-Type: text/plain; charset=utf-8

post /%
$ curl -i 'localhost:3232/query?lang=go'
HTTP/1.1 200 OK
Date: Wed, 29 Mar 2017 06:35:55 GMT
Content-Length: 17
Content-Type: text/plain; charset=utf-8

has lang=go query%
$ curl -i localhost:3232/headers
HTTP/1.1 200 OK
X-Api-Faker: true
X-Lang: Go
Date: Wed, 29 Mar 2017 06:37:14 GMT
Content-Length: 27
Content-Type: text/plain; charset=utf-8

respond with custom headers%
$ curl -i localhost:3232/json
HTTP/1.1 200 OK
Content-Type: application/json
Date: Wed, 29 Mar 2017 06:37:35 GMT
Content-Length: 50

{
  "name": "api-faker",
  "msg": "hello world"
}
$ curl -i localhost:32323/created
HTTP/1.1 201 Created
Date: Wed, 29 Mar 2017 06:37:48 GMT
Content-Length: 7
Content-Type: text/plain; charset=utf-8

created%
```



CHANGELOG.md
# Change Log

## 0.1.1 - 2017.4.12

- 配置项的`file`路径应该是相对于配置文件的

## 0.1.0 - 2017.3.29

- 基础版本


main.go
/*
* @Author: CJ Ting
* @Date:   2017-03-28 18:22:14
* @Last Modified by:   CJ Ting
* @Last Modified time: 2017-04-12 18:53:35
 */

// API Faker creates a server based on a yaml config file
package main

import (
	"io/ioutil"
	"mime"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	"gopkg.in/alecthomas/kingpin.v2"

	log "github.com/Sirupsen/logrus"
	"gopkg.in/yaml.v2"
)

const watchDuration = 500 * time.Millisecond

var (
	port       int
	configPath string

	// protect items
	mutex sync.Mutex
	items []*Item
)

var appVersion string // set by -ldflags

var configDir string // config file dir

type Item struct {
	Path string
	// default value: "GET"
	Method  string
	Body    string
	Headers map[string]string
	Query   map[string]string
	// default value: 200
	Code int
	File string
}

func (item *Item) Matched(path string, method string, query url.Values) bool {
	if item.Path != path {
		return false
	}

	if item.getMethod() != method {
		return false
	}

	if item.Query != nil {
		for k, v := range item.Query {
			if query.Get(k) != v {
				return false
			}
		}
	}
	return true
}

func (item *Item) getMethod() string {
	m := strings.ToUpper(item.Method)
	if m == "" {
		m = "GET"
	}
	return m
}

func (item *Item) getCode() int {
	c := item.Code
	if c == 0 {
		c = 200
	}
	return c
}

func init() {
	log.SetFormatter(&log.TextFormatter{
		FullTimestamp: true,
	})
}

func parseFlags() {
	kingpin.Flag("port", "listening port").Default("3232").IntVar(&port)
	kingpin.Version(appVersion)
	kingpin.CommandLine.HelpFlag.Short('h')

	kingpin.Arg("config", "config file").Required().StringVar(&configPath)

	kingpin.Parse()
}

func parseConfig(buf []byte) {
	mutex.Lock()
	defer mutex.Unlock()
	if err := yaml.Unmarshal(buf, &items); err != nil {
		log.WithError(err).Fatal("Failed to parse config")
	} else {
		log.Info("Parse config successfully")
	}
}

func main() {
	parseFlags()

	buf, err := ioutil.ReadFile(configPath)
	if err != nil {
		log.WithError(err).
			WithField("config", configPath).
			Fatal("Failed to read file")
	}

	configAbsPath, err := filepath.Abs(configPath)
	if err != nil {
		log.WithError(err).Fatal("Failed to get config abs path")
	}
	configDir = filepath.Dir(configAbsPath)

	parseConfig(buf)

	// watch config file
	go watchConfig()

	server := createServer()

	log.WithField("port", port).Info("Server started")

	log.WithError(http.ListenAndServe(":"+strconv.Itoa(port), server)).Fatal("Server crashed")
}

func createServer() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		status := http.StatusOK
		url := r.URL.String()
		path := r.URL.Path

		// log
		defer func() {
			log.WithFields(log.Fields{
				"method": r.Method,
				"code":   status,
			}).Info(url)
		}()

		var target *Item

		mutex.Lock()
		for _, item := range items {
			if item.Matched(path, r.Method, r.URL.Query()) {
				target = item
				break
			}
		}
		mutex.Unlock()

		if target == nil {
			status = http.StatusNotFound
			w.WriteHeader(status)
			return
		}

		status = target.getCode()

		// write headers
		for k, v := range target.Headers {
			w.Header().Set(k, v)
		}

		// set MIME content type
		if target.File != "" {
			mime := mime.TypeByExtension(filepath.Ext(target.File))
			w.Header().Set("Content-Type", mime)
		}

		var buf []byte

		// write body
		if target.File != "" {
			var err error
			filePath := target.File
			if !filepath.IsAbs(filePath) {
				filePath = filepath.Join(configDir, filePath)
			}
			buf, err = ioutil.ReadFile(filePath)
			if err != nil {
				log.WithError(err).WithFields(log.Fields{
					"url":  url,
					"file": target.File,
				}).
					Error("Failed to read file")
				status = http.StatusInternalServerError
			}
		} else {
			buf = []byte(target.Body)
		}

		w.WriteHeader(status)
		w.Write(buf)
	})
}

func watchConfig() {
	lastModTime := time.Now()
	for {
		stat, err := os.Stat(configPath)
		if err != nil {
			log.WithError(err).WithField("config", configPath).Error("Can't get config stat")
		} else {
			if stat.ModTime().After(lastModTime) {
				lastModTime = stat.ModTime()
				log.Info("Detect config file updated, reparse...")
				buf, err := ioutil.ReadFile(configPath)
				if err != nil {
					log.WithError(err).WithField("config", configPath).Error("Can't read config file")
				} else {
					parseConfig(buf)
				}
			}
		}

		time.Sleep(watchDuration)
	}
}
